File: algorithms/code/leetcode/lc020_valid_parentheses/LC020ValidParentheses.java
package leetcode.lc020_valid_parentheses;

import java.util.Stack;

/**
 * https://leetcode.com/problems/valid-parentheses/
 */
public final class LC020ValidParentheses {
    public boolean isValid(final String s) {
        if (s == null) {
            return false;
        }
        if (s.length() == 0) {
            return true;
        }
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char character = s.charAt(i);
            switch (character) {
                case '(':
                case '{':
                case '[':
                    stack.push(character);
                    break;
                case ')':
                    if (stack.size() == 0 || stack.peek() != '(') {
                        return false;
                    } else {
                        stack.pop();
                    }
                    break;
                case '}':
                    if (stack.size() == 0 || stack.peek() != '{') {
                        return false;
                    } else {
                        stack.pop();
                    }
                    break;
                case ']':
                    if (stack.size() == 0 || stack.peek() != '[') {
                        return false;
                    } else {
                        stack.pop();
                    }
                    break;
                default:
                    return false;
            }
        }
        return stack.size() == 0;
    }
}

File: algorithms/code/leetcode/lc020_valid_parentheses/LC020ValidParentheses.cs
// https://leetcode.com/problems/valid-parentheses/

using System.Collections.Generic;

namespace AlgorithmDesign.code.leetcode.lc020_valid_parentheses
{
    public class Solution
    {
        public bool IsValid(string s)
        {
            if (s == null) return false;
            if (s.Length == 0) return true;
            Stack<char> stack = new Stack<char>();
            foreach (char character in s)
            {
                switch (character)
                {
                    case '(':
                    case '{':
                    case '[':
                        stack.Push(character);
                        break;
                    case ')':
                        if (stack.Count == 0 || stack.Peek() != '(')
                        {
                            return false;
                        }
                        stack.Pop();
                        break;
                    case '}':
                        if (stack.Count == 0 || stack.Peek() != '{')
                        {
                            return false;
                        }
                        stack.Pop();
                        break;
                    case ']':
                        if (stack.Count == 0 || stack.Peek() != '[')
                        {
                            return false;
                        }
                        stack.Pop();
                        break;
                    default:
                        return false;
                }
            }
            return stack.Count == 0;
        }
    }
}
